Content.Server/Anomaly/AnomalySystem.Behaviour.cs

```diff
@@ -0,0 +1,104 @@
+    }
+
+    private void OnPrototypesReloaded(PrototypesReloadedEventArgs ev)
+    {
+        _behaviourList.Clear();
+        _behaviourList.AddRange(_prototype.EnumeratePrototypes<AnomalyBehaviourPrototype>());
+    }
+
+    private string GetRandomBehaviour()
+    {
+        var totalWeight = _behaviourList.Sum(x => x.Weight);
+        var randomValue = _random.NextFloat(totalWeight);
+
+        foreach (var b in _behaviourList)
+        {
+            if (randomValue < b.Weight)
+            {
+                return b.ID;
+            }
+
+            randomValue -= b.Weight;
+        }
+        // Shouldn't happen
+        throw new InvalidOperationException($"Invalid weighted variantize anomaly behaviour pick!");
+    }
+
+    private void SetBehaviour(Entity<AnomalyComponent> anomaly, ProtoId<AnomalyBehaviourPrototype> behaviourProto)
+    {
+        if (anomaly.Comp.CurrentBehaviour == behaviourProto)
+            return;
+
+        if (anomaly.Comp.CurrentBehaviour != null)
+            RemoveBehaviour(anomaly);
+
+        anomaly.Comp.CurrentBehaviour = behaviourProto;
+
+        var behaviour = _prototype.Index(behaviourProto);
+
+        foreach (var (name, entry) in behaviour.Components)
+        {
+            var reg = _componentFactory.GetRegistration(name);
+
+            if (EntityManager.HasComponent(anomaly, reg.Type))
+            {
+                EntityManager.RemoveComponent(anomaly, reg.Type);
+            }
+
+            var comp = (Component) _componentFactory.GetComponent(reg);
+            comp.Owner = anomaly;
+
+            var temp = (object) comp;
+            _serialization.CopyTo(entry.Component, ref temp);
+            EntityManager.RemoveComponent(anomaly, temp!.GetType());
+            EntityManager.AddComponent(anomaly, (Component) temp!);
+        }
+    }
+
+    private void RemoveBehaviour(Entity<AnomalyComponent> anomaly)
+    {
+        if (anomaly.Comp.CurrentBehaviour == null)
+            return;
+
+        var behaviour = _prototype.Index(anomaly.Comp.CurrentBehaviour.Value);
+
+        var entityPrototype = MetaData(anomaly).EntityPrototype;
+        var toRemove = behaviour.Components.Keys.ToList();
+
+        foreach (var name in toRemove)
+        {
+            // if the entity prototype contained the component originally
+            if (entityPrototype?.Components.TryGetComponent(name, out var entry) ?? false)
+            {
+                var comp = (Component) _componentFactory.GetComponent(name);
+                comp.Owner = anomaly;
+                var temp = (object) comp;
+                _serialization.CopyTo(entry, ref temp);
+                EntityManager.RemoveComponent(anomaly, temp!.GetType());
+                EntityManager.AddComponent(anomaly, (Component) temp);
+                continue;
+            }
+
+            EntityManager.RemoveComponentDeferred(anomaly, _componentFactory.GetRegistration(name).Type);
+        }
```

Comment:
Prototypes should only be used for spawning and nothing more so this needs fixing, also you've copied this code a bunch which you should add it to a common area if you really want.

Reply:
need help with it